okhttp/src/main/java/okhttp3/internal/publicsuffix/PublicSuffixDatabase.java

```diff
@@ -0,0 +1,325 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal.publicsuffix;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.net.IDN;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicBoolean;
+import okhttp3.internal.Util;
+import okhttp3.internal.platform.Platform;
+import okio.BufferedSource;
+import okio.GzipSource;
+import okio.Okio;
+
+import static okhttp3.internal.Util.closeQuietly;
+
+/**
+ * A database of public suffixes provided by
+ * <a href="https://publicsuffix.org/">publicsuffix.org</a>.
+ */
+public final class PublicSuffixDatabase {
+  public static final String PUBLIC_SUFFIX_RESOURCE = "publicsuffixes.gz";
+
+  private static final byte[] WILDCARD_LABEL = new byte[]{'*'};
+  private static final String[] EMPTY_RULE = new String[0];
+  private static final String[] PREVAILING_RULE = new String[]{"*"};
+
+  private static final byte EXCEPTION_MARKER = '!';
+
+  /** True after we've attempted to read the list for the first time. */
+  private final AtomicBoolean listRead = new AtomicBoolean(false);
+
+  /** Used for concurrent threads reading the list for the first time. */
+  private final CountDownLatch readCompleteLatch = new CountDownLatch(1);
+
+  // The lists are held as a large array of UTF-8 bytes. This is to avoid allocating lots of strings
+  // that will likely never be used. Each rule is separated by '\n'. Please see the
+  // PublicSuffixListGenerator class for how these lists are generated.
+  // Guarded by this.
+  private byte[] publicSuffixListBytes;
+  private byte[] publicSuffixExceptionListBytes;
+
+  /**
+   * Returns the effective top-level domain plus one (eTLD+1) by referencing the public suffix list.
+   * Returns null if the domain is a public suffix or the list was unable to be loaded.
+   *
+   * <p>Here are some examples: <pre>{@code
+   * assertEquals("google.com", getEffectiveTldPlusOne("google.com"));
+   * assertEquals("google.com", getEffectiveTldPlusOne("www.google.com"));
+   * assertNull(getEffectiveTldPlusOne("com"));
+   * }</pre>
+   */
+  public String getEffectiveTldPlusOne(String domain) {
+    if (domain == null) return null;
+
+    // Confirm we're dealing with a valid domain.
+    // TODO: Domains that start with '.' will be null on JDK8, but not on JDK7.
+    String canonicalDomain = Util.domainToAscii(domain);
+    if (canonicalDomain == null || canonicalDomain.startsWith(".")) return null;
+
+    // Convert it back to Unicode because the list uses Unicode, not punycode form.
+    String unicodeDomain = IDN.toUnicode(canonicalDomain);
+
+    // Preserve the passed in domain format (Unicode vs. punycode), but canonicalize it. This
+    // satisfies the publicsuffix.org test cases, but it's unclear if this should be done generally.
+    String domainToReturn = canonicalDomain.equalsIgnoreCase(domain)
```

Comment:
👍 - that will get rid of this madness. It'll mean tweaking the publicsuffix.org test cases slightly. I'll follow-up with that.